### ISSUE TITLE
Where Union of Labels

### ISSUE THREAD

**Issue by jtomek** [apache/age#1714]
Hi, how do I run this?
```sql
SELECT * FROM cypher('playground', $$
MATCH (n:Movie|Person)
RETURN n
$$) AS (v agtype)
```

I get 

```markdown
Query 1 ERROR at Line 31: : ERROR:  syntax error at or near "|"
LINE 2: MATCH (n:Movie|Person)
                      ^
```

Is this the only way to get what I want?

```sql
SELECT * FROM cypher('playground', $$
MATCH (n)
WHERE label(n) = 'Movie' or label(n) = 'Person'
RETURN n
$$) AS (v agtype)
```sql

**Comment by rafsun42** [comment]
@jtomek There is a PR (#1452) for this feature. It is based on PG15.

**Comment by diangamichael** [comment]
It seems you're trying to execute Cypher queries within a PostgreSQL environment using the cypher function. The error you're encountering is due to the incorrect usage of the '|' symbol in the label definition within your MATCH clause. In Cypher, '|' is not used to denote multiple labels for a node. Instead, you should use the IN keyword.

Here's the corrected version of your first query:

sql
Copy code
SELECT * FROM cypher('playground', $$
MATCH (n)
WHERE n:Movie OR n:Person
RETURN n
$$) AS (v agtype);
This query will match nodes that have either the label 'Movie' or 'Person'.

Alternatively, you can use your second query as you provided:

sql
Copy code
SELECT * FROM cypher('playground', $$
MATCH (n)
WHERE label(n) = 'Movie' OR label(n) = 'Person'
RETURN n
$$) AS (v agtype);

Both queries should achieve the same result, with the second one explicitly checking the labels of nodes. Choose whichever you find more readable or preferable for your use case.

**Comment by rafsun42** [comment]
@diangamichael 

> In Cypher, '|' is not used to denote multiple labels for a node.

This is not correct. According to the openCypher specification, the pipe operator '|' is a valid way to define union of multiple labels.

**Comment by alexgraul** [comment]
`WHERE label(n) = 'Movie' OR label(n) = 'Person'` works but is fairly ungainly. Per the openCypher spec `(:Foo|Bar)` should be supported.

**Comment by jrgemignani** [comment]
Apache AGE does **not** support multiple labels, yet. There is a PR **in review** that could add this in the future. However, it is a large PR and needs to be thoroughly reviewed.

**Comment by alexgraul** [comment]
Hey @jrgemignani thankyou for chiming in, aware of the multi-label PR but this is a separate issue with querying, effectively being able to OR between labels in a query. ie if you have the following in your db
```
(:City)-[:AirRoute]-(:City)
```

**Comment by alexgraul** [comment]
Hey @jrgemignani thankyou for chiming in, aware of the multi-label PR but this is a separate issue with querying, effectively being able to OR between labels in a query which is relevant for single labels, not ANDing for situations where you want to match on multiple labels on a query which I believe in openCypher terms is expressed as `[:Foo:Bar]`.
So if you have the following in your db
```
(:City)-[:AirRoute]-(:City)
(:City)-[:SeaRoute]-(:City)
```
Being able to run a query and say 'Give me any Air OR Sea route between these two cities', which should be possible (in openCypher terms) via 
```
(:City { name: 'London' })-[:SeaRoute|AirRoute]-(:City { name: :'Rotterdam' })
```
You can do it today via a WHERE clause as shown above but the syntax is super clunky.

**Comment by rafsun42** [comment]
Hi @alexgraul Although the PR refers it as 'multi-label', it does support the `|` operator for combining multiple single labels in MATCH queries. For now you can look at the [regression test file](https://github.com/apache/age/blob/0a37397ce6c958a4deb7c9d710b64bfcf1af0876/regress/sql/multiple_label.sql) to see what kind of multi-label queries are supported.

**Comment by alexgraul** [comment]
ohhh nice! Awesome, thankyou for clarifying!

**Comment by Mocuto** [comment]
Leaving a comment here to ask for any updates on this issue. I see #2082 has been approved. The lack of union support, combined with lack of an `all` predicate makes querying for paths with edge labels within a specific subset very difficult.  To use a modification of the above example, you might be interested in any variable length route between two cities that uses either air or sea routes:

```
path=(:City { name: 'London' })-[:SeaRoute|AirRoute*1..]->(:City { name: :'Rotterdam' })
```

This sort of path query works fine with Neo4j but I can't figure out a workaround with Apache AGE.
